chain/substreams/src/trigger.rs

```diff
@@ -85,3 +100,133 @@ impl blockchain::TriggersAdapter<Chain> for TriggersAdapter {
         unimplemented!()
     }
 }
+
+fn write_poi_event(
+    proof_of_indexing: &SharedProofOfIndexing,
+    poi_event: &ProofOfIndexingEvent,
+    causality_region: &str,
+    logger: &Logger,
+) {
+    if let Some(proof_of_indexing) = proof_of_indexing {
+        let mut proof_of_indexing = proof_of_indexing.deref().borrow_mut();
+        proof_of_indexing.write(logger, causality_region, poi_event);
+    }
+}
+
+pub struct TriggerProcessor {
+    pub locator: DeploymentLocator,
+}
+
+impl TriggerProcessor {
+    pub fn new(locator: DeploymentLocator) -> Self {
+        Self { locator }
+    }
+}
+
+#[async_trait]
+impl<T> graph_core::TriggerProcessor<Chain, T> for TriggerProcessor
+where
+    T: RuntimeHostBuilder<Chain>,
+{
+    async fn process_trigger(
+        &self,
+        logger: &Logger,
+        _hosts: &[Arc<T::Host>],
+        block: &Arc<Block>,
+        _trigger: &TriggerData,
+        mut state: BlockState<Chain>,
+        proof_of_indexing: &SharedProofOfIndexing,
+        causality_region: &str,
+        _debug_fork: &Option<Arc<dyn SubgraphFork>>,
+        _subgraph_metrics: &Arc<SubgraphInstanceMetrics>,
+    ) -> Result<BlockState<Chain>, MappingError> {
+        for entity_change in block.entities_changes.entity_changes.iter() {
+            match entity_change.operation() {
+                Operation::Unset => {
+                    // Potentially an issue with the server side or
+                    // we are running an outdated version. In either case we should abort.
+                    return Err(MappingError::Unknown(anyhow!("Detected UNSET entity operation, either a server error or there's a new type of operation and we're running an outdated protobuf")));
+                }
+                Operation::Create | Operation::Update => {
+                    let entity_type: &str = &entity_change.entity;
+                    let entity_id: String = String::from_utf8(entity_change.id.clone())
```

Comment:
I don't see why `clone` would be required here.

Reply:
entity_change is a reference and so the vec can't be moved.

Reply:
But `String::from_utf8` is accepting a reference of bytes `&[u8]` creating a new `String` instance, I don't see `String::from_utf8` owning the buffer so I don't see why the `clone` is necessary.

Reply:
Scratch all that! I was looking at `str::from_utf8` sorry.